contracts/artifacts/DelegatedOptIn.approval.teal

```diff
@@ -0,0 +1,530 @@
+	frame_dig -2 // preMbr: uint64
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	acct_params_get AcctMinBalance
+	assert
+	-
+	frame_bury -3 // mbrDelta: uint64
+
+	// ./contracts/delegated_optin_app.algo.ts:76
+	// assert(mbrDelta !== this.assetMBR.get())
+	frame_dig -3 // mbrDelta: uint64
+	byte "assetMBR"
+	app_global_get
+	!=
+	assert
+
+	// ./contracts/delegated_optin_app.algo.ts:77
+	// this.assetMBR.set(mbrDelta)
+	byte "assetMBR"
+	frame_dig -3 // mbrDelta: uint64
+	app_global_put
+
+	// ./contracts/delegated_optin_app.algo.ts:79
+	// sendAssetTransfer({
+	itxn_begin
+	int axfer
+	itxn_field TypeEnum
+
+	// ./contracts/delegated_optin_app.algo.ts:80
+	// assetReceiver: this.app.address
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	itxn_field AssetReceiver
+
+	// ./contracts/delegated_optin_app.algo.ts:81
+	// xferAsset: asset
+	frame_dig -1 // asset: asset
+	itxn_field XferAsset
+
+	// ./contracts/delegated_optin_app.algo.ts:82
+	// assetAmount: 0
+	int 0
+	itxn_field AssetAmount
+
+	// ./contracts/delegated_optin_app.algo.ts:83
+	// fee: 0
+	int 0
+	itxn_field Fee
+
+	// ./contracts/delegated_optin_app.algo.ts:84
+	// assetCloseTo: this.app.address
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	itxn_field AssetCloseTo
+	itxn_submit
+	retsub
+
+abi_route_setOpenOptInSignature:
+	txn OnCompletion
+	int NoOp
+	==
+	txn ApplicationID
+	int 0
+	!=
+	&&
+	assert
+
+	// no dupn needed
+	txn GroupIndex
+	int 1
+	-
+	txna ApplicationArgs 2
+	txna ApplicationArgs 1
+	callsub setOpenOptInSignature
+	int 1
+	return
+
+setOpenOptInSignature:
+	proto 3 0
+
+	// ./contracts/delegated_optin_app.algo.ts:99
+	// assert(verifier.sender === this.sigVerificationAddress.get())
+	frame_dig -3 // verifier: txn
+	gtxns Sender
+	byte "sigVerificationAddress"
+	app_global_get
+	==
+	assert
+
+	// ./contracts/delegated_optin_app.algo.ts:101
+	// this.openOptInSignatures.set(signer, sig)
+	byte "s-"
+	frame_dig -2 // signer: address
+	concat
+	frame_dig -1 // sig: byte[64]
+	box_put
+	retsub
+
+abi_route_openOptIn:
+	txn OnCompletion
+	int NoOp
+	==
+	txn ApplicationID
+	int 0
+	!=
+	&&
+	assert
+
+	// no dupn needed
+	txn GroupIndex
+	int 1
+	-
+	txn GroupIndex
+	int 2
+	-
+	callsub openOptIn
+	int 1
+	return
+
+openOptIn:
+	proto 2 0
+
+	// ./contracts/delegated_optin_app.algo.ts:113
+	// assert(optIn.assetReceiver === mbrPayment.receiver)
+	frame_dig -2 // optIn: axfer
+	gtxns AssetReceiver
+	frame_dig -1 // mbrPayment: pay
+	gtxns Receiver
+	==
+	assert
+
+	// ./contracts/delegated_optin_app.algo.ts:114
+	// assert(mbrPayment.amount >= this.assetMBR.get())
+	frame_dig -1 // mbrPayment: pay
+	gtxns Amount
+	byte "assetMBR"
+	app_global_get
+	>=
+	assert
+
+	// if0_condition
+	// ./contracts/delegated_optin_app.algo.ts:117
+	// this.openOptInEndTimes.exists(optIn.assetReceiver)
+	byte "e-"
+	frame_dig -2 // optIn: axfer
+	gtxns AssetReceiver
+	concat
+	box_len
+	swap
+	pop
+	bz if0_end
```

Comment:
The TEAL is generated by TEALScript, which has seperate `get` and `exist` functions and no direct exposure of `box_get` without popping one of the values. Compiler optimization can eventually be made to account for this sort of pattern but in the meantime the TEAL would need to be manually edited to optimize this.

Reply:
This isn't the only place where this extra complexity is coming up.  If this is going to be audited, they are going to have to audit the TEAL, as they are not going to be able to trust all of TEALScipt compilation without auditing _it_.

There are a lot of places where we're creating a lot more audit work by auditing compiler output. (things like confirming that each methods checks `OnCompletion` instead of doing it in one place, the exxtra `callsubs` and "trampolines" for each method, etc.

Reply:
I question how much extra "audit work" an extra `box_len` or `callsub` really adds. I'm not super familiar with the auditing process, but I know we got a reasonable quote and timeline for ARC12 which was TEALScript-generated and more complex.

Regarding OnCompletion, it's been on the TODO list to have only one entry point per OnCompletion in the generated TEAL from TEALScript, so this is something that can be updated.